File: packages/api/src/novels/novels.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  NotFoundException,
  Param,
  Post,
  Query,
} from '@nestjs/common';
import { UpdateNovelStarDto } from './dto/update-novel-star.dto';
import { NovelsService } from './novels.service';

@Controller()
export class NovelsController {
  constructor(private readonly novels: NovelsService) {}

  @Get('/novels')
  getNovels(
    @Query('page') page = 1,
    @Query('limit') limit = 1000,
  ) {
    return this.novels.getNovelsLimit(Number(page), Number(limit));
  }

  @Get('/novelByName/:name')
  async getNovelByName(@Param('name') name: string) {
    const novel = await this.novels.getNovelByName(decodeURIComponent(name));
    if (!novel) throw new NotFoundException('Novel not found');
    return novel;
  }

  @Get('/novel/:id')
  getNovel(@Param('id') id: string, @Query('page') page?: number) {
    return this.novels.getNovelPage(Number(id), page);
  }

  @Post('/novel/:id')
  updateStartRating(
    @Param('id') id: string,
    @Body() body: UpdateNovelStarDto,
  ) {
    return this.novels.updateStarRating(Number(id), body.starRating);
  }

  @Get('/scanning')
  doScanning() {
    return this.novels.doScanning();
  }

  @Delete('/novel/:id')
  async deleteNovel(@Param('id') id: string) {
    try {
      await this.novels.deleteNovel(Number(id));
      return { affected: 1, raw: [] as unknown[] };
    } catch {
      return { affected: 0, raw: [] as unknown[] };
    }
  }
}

File: packages/java/src/main/java/com/videomanager/novels/NovelsController.java
package com.videomanager.novels;

import com.videomanager.novels.dto.UpdateNovelStarDto;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class NovelsController {

    private final NovelsService novelsService;

    public NovelsController(NovelsService novelsService) {
        this.novelsService = novelsService;
    }

    @GetMapping("/novels")
    public Object getNovels(
        @RequestParam(defaultValue = "1") int page,
        @RequestParam(defaultValue = "1000") int limit
    ) {
        return novelsService.getNovelsLimit(page, limit);
    }

    @GetMapping("/novelByName/{name}")
    public Object getNovelByName(@PathVariable String name) {
        return novelsService.getNovelByName(name);
    }

    @GetMapping("/novel/{id}")
    public Object getNovel(@PathVariable int id, @RequestParam(required = false) Integer page) {
        return novelsService.getNovelPage(id, page);
    }

    @PostMapping("/novel/{id}")
    public Object updateStarRating(
        @PathVariable int id,
        @Valid @RequestBody UpdateNovelStarDto dto
    ) {
        return novelsService.updateStarRating(id, dto.starRating());
    }

    @GetMapping("/scanning")
    public Object scanning() {
        return novelsService.doScanning();
    }

    @DeleteMapping("/novel/{id}")
    public Object deleteNovel(@PathVariable int id) {
        return novelsService.deleteNovel(id);
    }
}
